app/router/profile.py

```diff
@@ -401,14 +401,71 @@ def get_coach_cards(
         coach_ids = [cs.coach_id for cs in coach_sports]
         query = query.filter(m.Coach.id.in_(coach_ids))
     if address:
-        # main filtration # TODO
-        # search by numbers if postal code
-        location = (
-            db.query(m.Location).filter(m.Location.city.icontains(f"{address}")).first()
-        )
-        coach_ids = [
-            cl.coach_id
-            for cl in db.query(m.CoachLocation).filter_by(location_id=location.id).all()
-        ]
+        if re.match(r"^([\s\d]+)$", address):
+            locations = (
+                db.query(m.Location)
+                .filter(
+                    or_(
+                        m.Location.postal_code.icontains(f"{address}"),
+                    )
+                )
+                .all()
+            )
+
+        elif " " in address:
+            address_line1 = address.split(" ")[0]
+            address_line2 = address.split(" ")[1]
+            locations = (
+                db.query(m.Location)
+                .filter(
+                    or_(
+                        and_(
+                            m.Location.city.icontains(f"{address_line1}"),
+                            m.Location.street.icontains(f"{address_line2}"),
+                        ),
+                        # e.g. we have multiple words in CITY name
+                        and_(
+                            m.Location.city.icontains(
+                                f"{address_line1} {address_line2}"
+                            ),
+                        ),
+                        # e.g. we have multiple words in STREET name
+                        and_(
+                            m.Location.street.icontains(
+                                f"{address_line1} {address_line2}"
+                            ),
+                        ),
+                        and_(
+                            m.Location.city.icontains(f"{address_line1}"),
+                            m.Location.street.icontains(f"{address_line1}"),
+                        ),
+                    )
+                )
+                .all()
+            )
+
+        else:
+            locations = (
+                db.query(m.Location)
+                .filter(
+                    or_(
+                        m.Location.city.icontains(f"{address}"),
+                        m.Location.street.icontains(f"{address}"),
+                    )
+                )
+                .all()
+            )
+
+        location_ids = [location.id for location in locations]
+        if not location_ids:
+            coach_ids = []
+        else:
+            coach_ids = [
+                cl.coach_id
+                for cl in db.query(m.CoachLocation)
+                .filter(m.CoachLocation.location_id.in_(location_ids))
+                .all()
+            ]
+
```

Comment:
As for me, this function is too big and has too many complicated nestings. I can only assume it does what it has to do. I'd suggest separate this kind of logic into a bunch of separate small functions at least (basically convert each logic block under `if/elif/else` into a separate function. At this point the overall function logic may become more clear